pilot/pkg/config/coredatamodel/incrementalController.go

```diff
@@ -0,0 +1,374 @@
+
+	for _, fullName := range configName {
+		namespace, name := extractNameNamespace(fullName)
+		if byNamespace, ok := c.configStore[namespace]; ok {
+			if conf, ok := byNamespace[name]; ok {
+				delete(byNamespace, conf.Name)
+			}
+			// clear parent map also
+			if len(byNamespace) == 0 {
+				delete(byNamespace, namespace)
+			}
+		}
+	}
+}
+
+func (c *IncrementalController) convertToConfig(obj *sink.Object) (conf *model.Config, err error) {
+	namespace, name := extractNameNamespace(obj.Metadata.Name)
+
+	createTime := time.Now()
+	if obj.Metadata.CreateTime != nil {
+		if createTime, err = types.TimestampFromProto(obj.Metadata.CreateTime); err != nil {
+			log.Warnf("Discarding incoming MCP resource: invalid resource timestamp (%s/%s): %v", namespace, name, err)
+			return nil, err
+		}
+	}
+
+	conf = &model.Config{
+		ConfigMeta: model.ConfigMeta{
+			Type:              model.SyntheticServiceEntry.Type,
+			Group:             model.SyntheticServiceEntry.Group,
+			Version:           model.SyntheticServiceEntry.Version,
+			Name:              name,
+			Namespace:         namespace,
+			ResourceVersion:   obj.Metadata.Version,
+			CreationTimestamp: createTime,
+			Labels:            obj.Metadata.Labels,
+			Annotations:       obj.Metadata.Annotations,
+			Domain:            c.options.DomainSuffix,
+		},
+		Spec: obj.Body,
+	}
+
+	schema, _ := c.ConfigDescriptor().GetByType(model.SyntheticServiceEntry.Type)
+	if err = schema.Validate(conf.Name, conf.Namespace, conf.Spec); err != nil {
+		log.Warnf("Discarding incoming MCP resource: validation failed (%s/%s): %v", conf.Namespace, conf.Name, err)
+		return nil, err
+	}
+	c.discovery.RegisterNotReadyEndpoints(conf)
+	return conf, nil
+
+}
+
+func (c *IncrementalController) configStoreUpdate(resources []*sink.Object) {
+	svcChanged := c.serviceChange(resources)
+	configs := make(map[string]map[string]*model.Config)
+	for _, obj := range resources {
+		conf, err := c.convertToConfig(obj)
+		if err != nil {
+			continue
+		}
+
+		namedConf, ok := configs[conf.Namespace]
+		if ok {
+			namedConf[conf.Name] = conf
+		} else {
+			configs[conf.Namespace] = map[string]*model.Config{
+				conf.Name: conf,
+			}
+		}
+
+		if svcChanged {
+			continue
+		}
+		// this is done before updating internal cache
+		oldEpVersion := c.endpointVersion(conf.Namespace, conf.Name)
+		newEpVersion := version(conf.Annotations, endpointKey)
+		if oldEpVersion != newEpVersion {
+			if err := c.edsUpdate(conf); err != nil {
+				log.Warnf("edsUpdate: %v", err)
+			}
+		}
+	}
+
+	c.configStoreMu.Lock()
+	c.configStore = configs
+	c.configStoreMu.Unlock()
+
+	if svcChanged {
+		c.options.XDSUpdater.ConfigUpdate(true)
+	}
+}
+
+func (c *IncrementalController) incrementalUpdate(resources []*sink.Object) {
+	svcChanged := c.serviceChange(resources)
+	for _, obj := range resources {
+		conf, err := c.convertToConfig(obj)
+		if err != nil {
+			continue
+		}
+
+		var oldEpVersion string
+		c.configStoreMu.Lock()
```

Comment:
Should you just lock on this once before the loop and then unlock immediately after?  You probably don't want to do the EDS updates with this lock, but you can just queue up the EDS updates and then execute them after you unlock.

Reply:
The EDSUpdate still takes place outside of the lock (I use the fresh config that I receive for EDSUpdate) the lock is only for updating the internal cache.